### PR TITLE
#1607: Added more tests to the kff package for better code coverage

### PR DESCRIPTION
**Pull request by andrewbp**
The work here aims to add unit tests to the kff test package to have better code coverage in the kff package. I am publishing this pull request as WIP to get feedback on progress thus far.

### CODE REVIEW

**Comment by jpdahlke** [comment]
Original kff package coverage:
```
Class: 86% 13/15    Method: 58% 93/158    Line: 52% 447/845
```
Updated kff package coverage:
```
Class: 100% 15/15   Method: 82% 130/158   Line: 73% 618/845
```

File: src/test/java/emissary/kff/KffQuincyFileTest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Override
+    @Before
+    public void setUp() throws Exception {
+        kffQuincyFile = new KffQuincyFile(resourcePath, "testFilter", KffFilter.FilterType.Unknown);
+    }
+
+    @Test
+    public void testCreation() {
+        assertEquals("MD5", kffQuincyFile.getPreferredAlgorithm());
```
Comment (jpdahlke) [inline src/test/java/emissary/kff/KffQuincyFileTest.java:25]: This should also assert recordLength is 16 as after the preferred alg, that's the reason for this object.

File: src/test/java/emissary/kff/KffFileTest.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    @Before
+    public void setUp() throws Exception {
+        kffFile = new KffFile(resourcePath, "testFilter", KffFilter.FilterType.Unknown);
+        kffFile.setPreferredAlgorithm("SHA-1");
+    }
+
+    @Override
+    @After
+    public void tearDown() throws Exception {
+        super.tearDown();
```
Comment (jpdahlke) [inline src/test/java/emissary/kff/KffFileTest.java:37]: Seems unnecessary?

**Comment by jpdahlke** [comment]
Good progress so far.

Some feedback:
- prefer using explicit static reference (ie: KffDataObjectHandler.KFF_PARAM_MD5 vs KFF_PARAM_MD5)
- avoid .* imports
- comb through the individual class breakouts when running coverage as there are still more low-hanging coverage pieces (ChecksumCalculator.getUseSsdeep()) that should be easy to include.
- the "main" test is a pretty neat and reasonable way of getting around the coverage problem for those methods 
- I prefer the static imports for assertions rather than Assert. usage.
- it might be worth using org.junit.jupiter for new test classes. we probably haven't been great doing that so far, but it's less classes to move later.
- I sure hope tmp.bin is a safe binary

File: src/test/java/emissary/kff/KffDataObjectHandlerTest.java [hunk 3]
```diff
@@ -51,4 +84,49 @@ public void testMapWithPrefix() {
...
+        assertArrayEquals(payload.data(), new byte[0]);
+    }
+
+    @Test
+    public void testParentToChildMethod() {
+        // String value = "value";
+        payload.setParameter(KFF_PARAM_DUPE_HIT, payload);
+        KffDataObjectHandler.parentToChild(payload);
+        assertNull(payload.getStringParameter(KFF_PARAM_DUPE_HIT));
+        // assertEquals(thing, parent.getParameter(KFF_PARAM_PARENT_HIT));
```
Comment (jpdahlke) [inline src/test/java/emissary/kff/KffDataObjectHandlerTest.java:110]: I'm assuming this is meant to be removed or reworked in an update.

**Review by jpdahlke — CHANGES_REQUESTED** [review]
I didn't do a full review of the WIP but a few things caught my eye.

**Comment by andrewbp** [comment]
Thanks for the feedback! I'll get to work on your requests.

File: src/test/java/emissary/kff/KffDataObjectHandlerTest.java (hunk 3 above)
Comment (andrewbp) [reply src/test/java/emissary/kff/KffDataObjectHandlerTest.java:110]: I can't recall what I was using this for or what I was going to do with this... I will remove it for now.

**Comment by andrewbp** [comment]
> * I sure hope tmp.bin is a safe binary

tmp.bin is derived from another bin that's used elsewhere for tests, so... it should be?

**Comment by andrewbp** [comment]
I believe I've addressed most of @jpdahlke concerns. KffResult could stand to use more coverage, but I'm putting out what I have so far.

**Comment by jpdahlke** [comment]
> > * I sure hope tmp.bin is a safe binary
> 
> I think it is? But how do I go about making sure?

Ha, well, where did it come from?

**Comment by jpdahlke** [comment]
> I believe I've addressed most of @jpdahlke concerns. KffResult could stand for more coverage, but I'm putting out what I have so far.
> 
> I wanted to note that you're going to see a lot of annoying diffs in the "assert" lines because jupiter changes the order of things. That's all that changed on those lines.

Sure, that's why I suggested doing it for the NEW classes.

**Comment by andrewbp** [comment]
> > > * I sure hope tmp.bin is a safe binary
> > 
> > 
> > I think it is? But how do I go about making sure?
> 
> Ha, well, where did it come from?

It was taken from and exists as a test file in our other codebase, as there is a similar test done there.

**Comment by jpdahlke** [comment]
> > > > * I sure hope tmp.bin is a safe binary
> > > 
> > > 
> > > I think it is? But how do I go about making sure?
> > 
> > 
> > Ha, well, where did it come from?
> 
> It was taken from and exists as a test file in our other codebase, as there is a similar test done there.

Ah, gotcha.

**Review by fbruton — CHANGES_REQUESTED** [review]
Replace assert.fail() and printStackTrace() calls with ```throw new AssertionError()```. See https://github.com/NationalSecurityAgency/emissary/pull/210/files

File: src/test/java/emissary/kff/KffFileTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Override
+    @Before
+    public void setUp() throws Exception {
+        kffFile = new KffFile(resourcePath, "testFilter", KffFilter.FilterType.Unknown);
+        kffFile.setPreferredAlgorithm("SHA-1");
+    }
+
+    @Test
+    public void testKffFleCreation() throws Exception {
```
[the remark this answers is not in the document]
Comment (jpdahlke) [reply src/test/java/emissary/kff/KffFileTest.java:32]: Well, since he's gonna snag that, check the spelling of "File" :)

**Comment by jpdahlke** [comment]
Sorry @andrewbp, I conflicted your pr w/ another. When you go to fix can you please rebase off the latest master before pushing?

**Comment by andrewbp** [comment]
All feedback has been addressed. I've been trying to properly squash the commits but the rebase is being weird.

**Review by fbruton — APPROVED** [review]

**Pull request merged**